gcp_variant_transforms/transforms/variant_to_bigquery.py

```diff
@@ -113,14 +116,34 @@ def expand(self, pcoll):
             self._allow_incompatible_records,
             self._omit_empty_sample_calls))
     if self._limited_write:
-      bq_rows |= 'LimitWrite' >> limit_write.LimitWrite(_WRITE_SHARDS_LIMIT)
-    return (bq_rows
-            | 'WriteToBigQuery' >> beam.io.Write(beam.io.BigQuerySink(
+      # We split data into _NUM_BQ_LIMITED_WRITE_PARTITIONS random partitions
+      # and then write each part to final BQ by appending them together.
+      # Combined with LimitWrite transform, this will avoid the BQ failure.
+      bq_row_partitions = bq_rows | beam.Partition(
+          lambda _, n: random.randint(0, n - 1),
+          _NUM_BQ_LIMITED_WRITE_PARTITIONS)
+      bq_writes = []
+      for i in range(_NUM_BQ_LIMITED_WRITE_PARTITIONS):
+        bq_rows = (bq_row_partitions[i] | 'LimitWrite' + str(i) >>
+                   limit_write.LimitWrite(_WRITE_SHARDS_LIMIT))
```

Comment:
It seems that we are doing both partitioning and GroupByKey (in LimitWrite), is this still needed? I mean do we have any evidence that both of these are required and partitioning/append alone is not enough?

Reply:
yes, both are still needed. Otherwise, you can still get more than 10K files per partition. We believe doing partition+limit write is the most effective approach and it speeds up the pipeline a lot (you take advantage of distributing the groupbykey step among more machines and it also runs 20 bq load operations in parallel rather than 1). We were able to successfully load 30B records to BQ in ~3h (all previous attempts failed). I think Saman also ran with 1000g as well.